cmd/generate/config/rules/sourcegraph.go

```diff
@@ -0,0 +1,29 @@
+package rules
+
+import (
+	"github.com/zricethezav/gitleaks/v8/cmd/generate/config/utils"
+	"github.com/zricethezav/gitleaks/v8/cmd/generate/secrets"
+	"github.com/zricethezav/gitleaks/v8/config"
+)
+
+func SourceGraph() *config.Rule {
+	// define rule
+	r := config.Rule{
+		RuleID:      "source-graph-token",
+		Description: "Sourcegraph is a code search and navigation engine.",
+		Regex:       utils.GenerateUniqueTokenRegex(`\b(sgp_(?:[a-fA-F0-9]{16}|local)_[a-fA-F0-9]{40}|sgp_[a-fA-F0-9]{40}|[a-fA-F0-9]{40})\b`, true),
+		Entropy:     3,
+		Keywords:    []string{"sgp_", "sourcegraph"},
+	}
+
+	// validate
+	tps := utils.GenerateSampleSecrets("sgp_", secrets.NewSecret(`\b(sgp_(?:[a-fA-F0-9]{16}|local)_[a-fA-F0-9]{40}|sgp_[a-fA-F0-9]{40}|[a-fA-F0-9]{40})\b`))
```

Comment:
We should add separate test cases for each pattern. This will only generate cases for a single pattern, meaning obvious issues could be missed.